### PR TITLE
[MM-40582] Send unsanitized user to source on user_updated

### PR DESCRIPTION
**Pull request by plant99**
#### Summary

On `user_update` event,  unsanitized user is sent to source of the event, to prevent extra ping to get user. Sanitized user data is sent to other users.

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-server/issues/19361

#### Release Note

```release-note
Added feature to send unsanitized user to source user on `user_updated` event.
```

### CODE REVIEW

**Comment by plant99** [comment]
cc: @agnivade 

Changes to webapp, mobile code are pending from my end. Meanwhile, please have a look if this logic looks right to you.

File: app/user.go [hunk 1]
```diff
@@ -1025,17 +1025,28 @@ func (a *App) UpdateUserAuth(userID string, userAuth *model.UserAuth) (*model.Us
...
+	omitUsers[user.Id] = true
 	a.SanitizeProfile(&user, false)
-	message := model.NewWebSocketEvent(model.WebsocketEventUserUpdated, "", "", "", nil)
+	message := model.NewWebSocketEvent(model.WebsocketEventUserUpdated, "", "", "", omitUsers)
 	message.Add("user", &user)
 	message.GetBroadcast().ContainsSanitizedData = true
 	a.Publish(message)
+
+	// send unsanitized user to event creator
+	sourceUserMessage := model.NewWebSocketEvent(model.WebsocketEventUserUpdated, "", "", unsanitizedCopyOfUser.Id, nil)
```
Comment (enahum) [inline app/user.go:1046]: correct me if I'm wrong, but I'm seeing 3 events being published here, shouldn't we ensure that the user only receives the one event.

What I see here is
`adminMessage` being broadcasted to everyone (this seems wrong, shouldn't this omit the source creator as well?)
`message` being broadcasted excluding the source creator (this looks ok to me)
`sourceUserMessage` being broadcasted only to the source creator (this looks ok to me)

Comment (plant99) [reply app/user.go:1046]: I think https://github.com/mattermost/mattermost-server/blob/master/app/web_conn.go#L721 forwards to all admins, so I get two events (one `sourceUserMessage ` and one `adminMessage`) when admin is the data source, thanks for pointing this out. I'll omit `sourceUser` from adminMessage.

Comment (plant99) [reply app/user.go:1049]: Please let me know if the latest commit fixes it?

Comment (agnivade) [reply app/user.go:1049]: Ah my apologies on that one. You are correct @enahum. Great catch!

File: app/web_conn.go [hunk 2]
```diff
@@ -732,7 +732,8 @@ func (wc *WebConn) shouldSendEvent(msg *model.WebSocketEvent) bool {
 			hasReadPrivateDataPermission = model.NewBool(wc.App.RolesGrantPermission(wc.GetSession().GetUserRoles(), model.PermissionManageSystem.Id))
 		}
 
-		if !*hasReadPrivateDataPermission {
+		// sensitive data should be allowed to be emitted back to source-user event.
+		if !*hasReadPrivateDataPermission && (wc.UserId != msg.GetBroadcast().UserId) {
```
Comment (agnivade) [inline app/web_conn.go:736]: We can revert this change. We only need to set `ContainsSanitizedData`, when we need to differentiate between users and admins. It's not perfectly made, but `ContainsSensitiveData` goes together with `ContainsSanitizedData` to separate users and admins.

Since we are directly sending the message to the user, we can ignore this part and it will match the condition in L742.

**Review by agnivade — COMMENTED** [review]
Looking great @plant99 ! Just some small changes needed.

File: app/web_conn.go (hunk 2 above)
Comment (plant99) [reply app/web_conn.go]: Understood the context, thanks @agnivade for the explanation. I've reverted these changes. Please review again?

**Comment by hmhealey** [comment]
@agnivade FYI, this is probably something that we just accept with this, but this could cause some weirdness with the cross-version cluster we've got on community. Theoretically, a user on community-daily could receive a santized version of themselves if they were updated by someone (themselves or an admin) on community. I don't think that would cause anything other than a temporary issue though, and I doubt it'll occur often anyway though, but I wanted to point out that it could happen in case someone else thinks up a way that could cause any other issues.

**Comment by agnivade** [comment]
The cross-version cluster is something that's going to go away soon. So I wouldn't be too worried about that. But yes, definitely a good point. :+1:

**Review by streamer45 — APPROVED** [review]
Looks great :+1:

**Comment by agnivade** [comment]
@hmhealey - Are we okay to merge, or you would prefer the webapp PR to be merged along with this? 0/5 from my side. LMK.

**Comment by hmhealey** [comment]
This is good to be merged first. It won't really have any effect on the front end until the web app changes are merged anyway.
